### PR TITLE
GH-572: As a developer I want async iteration (for await of loop)

### PR DESCRIPTION
**Pull request by mor-n4**
See #572.

### CODE REVIEW

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/AbstractProcessor.xtend [hunk 1]
```diff
@@ -86,67 +86,43 @@ package abstract class AbstractProcessor {
...
-					    val outerReturnTypeRef = TypeUtils.createGeneratorTypeRef(scope, funDef);
+					val needsRewrite =
+						(isAsync && !isGenerator && !TypeUtils.isPromise(innerReturnTypeRef, scope)) ||
+						(!isAsync && isGenerator && !TypeUtils.isGenerator(innerReturnTypeRef, scope)) ||
+						(isAsync && isGenerator && !TypeUtils.isAsyncGenerator(innerReturnTypeRef, scope));
+					if (needsRewrite) {
+					    val outerReturnTypeRef = if (!isGenerator) {
+					    	TypeUtils.createPromiseTypeRef(scope, innerReturnTypeRef, null);
+					    } else {
+						    TypeUtils.createGeneratorTypeRef(scope, funDef); // note: this method handles the choice Generator vs. AsyncGenerator
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/AbstractProcessor.xtend:124]: white space issue (mix of spaces and tabs)

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/utils/ExpectedTypeComputer.xtend [hunk 2]
```diff
@@ -136,25 +136,30 @@ package class ExpectedTypeComputer extends TypeSystemHelperStrategy {
 		G2.addThisType(myThisTypeRef); // takes the real-this type even if it is a type{this} reference.
 
 		if (funDef === null || !funDef.isGenerator)
-			return null; // yields only occur in generator functions
+			return null; // yield only occurs in generator functions
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/utils/ExpectedTypeComputer.xtend:139]: plural is correct

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/messages.properties [hunk 3]
```diff
@@ -765,12 +769,12 @@ KEY_THIS_REJECTED_IN_TOP_LEVEL_LAMBDA=error;;;In a top-level arrow function, the
...
-# no parameters required
-EXP_INDEXED_ACCESS_SYMBOL_WRONG_TYPE=error;;;Access of property Symbol.iterator only allowed for instances of Iterable and immediate instances of Object.
-# no parameters required
-EXP_INDEXED_ACCESS_SYMBOL_READONLY=error;;;Access to property Symbol.iterator of an Iterable is read-only.
+# 0: names of legal built-in symbol
+EXP_INDEXED_ACCESS_SYMBOL_INVALID=error;;;Indexed access with built-in symbols is only allowed for {0}.
+# 0: name of built-in symbol, 1: name of expected receiver type
+EXP_INDEXED_ACCESS_SYMBOL_WRONG_TYPE=error;;;Access of property Symbol.{0} only allowed for instances of {1}, immediate instances of Object, and dynamic types.
+# 0: name of built-in symbol, 1: name of receiver type
+EXP_INDEXED_ACCESS_SYMBOL_READONLY=error;;;Access to property Symbol.{0} of an {1} is read-only.
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/messages.properties:777]: 👍

**Review by mmews-n4 — APPROVED** [review]
👍 wow, quite a few changes. I also noticed some nice other minor clean-ups.

**Comment by mor-n4** [comment]
Thanks @mmews-n4!

Builds green, merging ...

**Pull request merged**